(Rebased) Font and draw_text adjustments

Splendide-Imaginarius opened the pull request:
Rebased version of https://github.com/mkxp-z/mkxp-z/pull/91 with a few additional fixes.

WaywardHeart committed:
Don't let drawn text move past the top of the rect

WaywardHeart committed:
Limit squeezing text to a minimum of 50%

Not quite accurate to what VX Ace is doing, but I think that's mostly because of kerning differences.

WaywardHeart committed:
Use the first version of the font we find

WaywardHeart committed:
Test for an invalid rect in drawText

WaywardHeart committed:
Match some outline positioning behavior from RGSS and enable transparent outlines

Outlines should 1) Not change the position of the text and 2) Be visible even if the text is cut off from reaching the edge of the rect, without leaving the rect itself.
We're currently not drawing full text for the outlines like RGSS does, so we don't perfectly succeed at (2), but if we ever start doing that then the logic is already there.

Text transparency with outlines present still doesn't match Enterbrain's implementation, but I gave up on trying to figure that out.

WaywardHeart committed:
Only draw enough of the text to fill the rect we were given

WaywardHeart committed:
Hack around an inconsistency in the width returned by Bitmap.text_size

freetype sometimes treats the last character of the string as being a pixel wider than it should be. Adding a space at the end and then removing it's width should make character-by-character text more accurate.

WaywardHeart committed:
Add config options to control how the font size is determined and toggle font kerning.

Some Pokemon Essentials games were explicitly made for mkxp. Text rendered using the old method tends to be smaller than what the new one produces, which means text in those games tends to get cut off with it.

Since Pokemon Essentials uses RGSS1, the setting defaults to the old method for RGSS1, and the new method for RGSS2 and RGSS3.

WaywardHeart committed:
Properly handle transparent text with outlines.

RGSS's outline is drawn by blitting a complete set of text four times at different offsets, which affects the final text color and opacity for transparent text. We need to replicate it.

The software blitting this commit introduces actually turned out to be faster than calling TTF_SetFontOutline, which apparently clears the font's glyph cache (which is *not* noted in SDL_ttf.h).

WaywardHeart committed:
Fix position of charLimit parameter in TTF_MeasureUTF8

WaywardHeart committed:
The default font size for VX Ace should be 24, not 22

WaywardHeart committed:
Close the RWops if the filesystem fails to open a font.

WaywardHeart committed:
Proper font sizing

Windows preferentially uses a special table to find the ppem for a requested height. Code taken from Wine's implementation.

WaywardHeart committed:
Go back to using SDL_ttf's outlines

Using RGSS's method looks very ugly in hires mode, and there's no easy way to fix it, so instead we need to fake the effect.

To speed up generating large menus, we're now also producing a second TTF_Font object for the outlines.

WaywardHeart committed:
Change how hires text sizing works.

Since the ppem doesn't directly correspond to the size now, it makes more sense to apply the hires multiplier to the ppem instead of the size. This commit's implementation has the side effect of not allowing drawing "lores" text directly to a hires bitmap; whatever size you set will always have the multiplier applied on top of it.

WaywardHeart committed:
I changed my mind about averaging the outline layers for transparent text color. The effect it was supposed to fake was only a small difference, anyway.

WaywardHeart committed:
Use the output of TTF_FontHeight for Bitmap#text_size and text positioning.

Splendide-Imaginarius committed:
Add test case for invalid draw_text rect

Splendide-Imaginarius committed:
Cite Wine font size code

Splendide-Imaginarius committed:
Add config option for font hinting

Splendide-Imaginarius committed:
Remove bug-for-bug compatibility with old mkxp font size calculation

Maintaining bug-for-bug compatibility with old mkxp versions is
impractical.

Splendide-Imaginarius committed:
Add config option for font height reporting

Splendide-Imaginarius committed:
The default font size for VX should be 20, not 22

Source: VX manual.

Splendide-Imaginarius committed:
Disable shadows if outlines are enabled

The shadows were being drawn with the text color instead of the shadow
color.

Splendide-Imaginarius committed:
Log BUG when drawing high-res text with shadow

Splendide-Imaginarius committed:
Add Eblo's draw_text postload test

Splendide-Imaginarius commented:
@Eblo would you be up for trying this out? I think it's ready to merge but it'd be nice to have someone besides me test it.

Splendide-Imaginarius commented:
Actually I just found what looks like a bug. It seems that with some fonts, the left edge of the leftmost glyph gets cropped off, which did not happen prior to this PR. (This is especially bad since a lot of games render text one glyph at a time.) I am not sure whether this happens in Enterbrain's runtime, I will try to reproduce it there.

Eblo commented:
Vanilla:
<img width="401" height="50" alt="image" src="https://github.com/user-attachments/assets/b30c5586-cbba-4603-9aef-0db8009d00a3" />

This fix:
<img width="401" height="50" alt="image" src="https://github.com/user-attachments/assets/e0ce7072-4371-4577-a869-a9d87482a924" />

This fix on top of vanilla:
<img width="401" height="50" alt="image" src="https://github.com/user-attachments/assets/4ce19427-c3a9-4614-b62e-4a83c2a67b8b" />

Looks pretty good. The glyph shapes seem to be aligned now. It seems that whatever vanilla is using seems to have softer color blending for the outlines, particularly with curves. Straight lines look more solidly black.

Splendide-Imaginarius commented:
@Eblo is that a blending difference or an opacity difference? E.g. if you set the outline opacity to 100% (default is 50% right?), do things look closer?

Eblo commented:
Seems like a blending difference. It's less stark if the opacity is 255, but there is still a difference around curves if you zoom in closely.

Comparisons (first two lines are default 128 opacity, second two lines are 255):

Vanilla:
<img width="408" height="120" alt="image" src="https://github.com/user-attachments/assets/7c9a3fe5-71c3-43d5-906c-c429ce3037cf" />

This fix:
<img width="408" height="120" alt="image" src="https://github.com/user-attachments/assets/53286e40-3c7f-4980-bda1-0900fbefacb3" />

This fix on top of vanilla:
<img width="408" height="120" alt="image" src="https://github.com/user-attachments/assets/ebca0e59-33ad-420f-a6cf-e51d6ad5ee15" />

Splendide-Imaginarius commented:
@white-axe You have access to big-endian hardware, right? Can you check if this PR breaks anything in BE environments? (There's some endian-dependent code in here that makes me nervous.)

white-axe commented:
I don't have big-endian hardware, but intuitively, the code seems fine. The new code in the blendText function in bitmap.cpp uses bit shifting to access the alpha component of a pixel, which will work regardless of endianness. The only other new endian-dependent code I see is the new GET_BE_WORD macro in font.cpp, but that macro is designed correctly for both big-endian and little-endian hardware.

Splendide-Imaginarius commented:
> Actually I just found what looks like a bug. It seems that with some fonts, the left edge of the leftmost glyph gets cropped off, which did not happen prior to this PR. (This is especially bad since a lot of games render text one glyph at a time.) I am not sure whether this happens in Enterbrain's runtime, I will try to reproduce it there.

The same bug happens in Enterbrain's runtime. I will add some test cases for this. Unfortunately I am reluctant to merge a PR that restores Enterbrain bugs without a way for users to disable that. So I will need to investigate how this can be worked around.

Splendide-Imaginarius committed:
Update text tests

Splendide-Imaginarius commented:
> Seems like a blending difference. It's less stark if the opacity is 255, but there is still a difference around curves if you zoom in closely.

@Eblo how about if you change the `fontHinting` config option? (I suspect that the "Light" mode might yield softer blending, although it will also probably change the geometry in ways that might or might not be a big deal.)

Splendide-Imaginarius committed:
Add config option for font outline cropping

Splendide-Imaginarius committed:
Support high-res text shadows

Splendide-Imaginarius commented:
> > Actually I just found what looks like a bug. It seems that with some fonts, the left edge of the leftmost glyph gets cropped off, which did not happen prior to this PR. (This is especially bad since a lot of games render text one glyph at a time.) I am not sure whether this happens in Enterbrain's runtime, I will try to reproduce it there.
> 
> The same bug happens in Enterbrain's runtime. I will add some test cases for this. Unfortunately I am reluctant to merge a PR that restores Enterbrain bugs without a way for users to disable that. So I will need to investigate how this can be worked around.

This turned out to be reasonably easy to fix; I've added a config option to toggle this Enterbrain bug. I also fixed high-res shadows.

At this point the only blocker for merging this PR is that shadows combined with outlines are broken. I will try to get that fixed in the next week or so.

Splendide-Imaginarius committed:
Support drawText with outline+shadow

Splendide-Imaginarius commented:
> At this point the only blocker for merging this PR is that shadows combined with outlines are broken. I will try to get that fixed in the next week or so.

Fixed. Wasn't terribly hard to deal with.

I did some further testing, and I see some oddities when the text color is translucent. I'd like to add some test cases for this so that I can make sure this is consistent with Enterbrain's behavior. I will try to do that in the next week.

Splendide-Imaginarius committed:
Add drawText translucency tests

Also add BUG debug statement for cases where it doesn't match Enterbrain
behavior.

Splendide-Imaginarius commented:
> I did some further testing, and I see some oddities when the text color is translucent. I'd like to add some test cases for this so that I can make sure this is consistent with Enterbrain's behavior. I will try to do that in the next week.

The color blending when outlines are enabled and the text color is non-opaque definitely does not match Enterbrain's behavior, but it's also a lot closer to Enterbrain's behavior with this PR than without this PR. So I'm not going to block this PR on that.

Splendide-Imaginarius commented:
The CI failures on cross-compiled Ubuntu seem to be unrelated to this PR (something to do with `apt` failing to install some packages).

Splendide-Imaginarius committed:
Fix memory safety errors in high-res text shadows

Splendide-Imaginarius commented:
I think this PR has been in incubation long enough, I'd like to merge it within the next 1-2 days. Last call for anyone who sees severe regressions to bring them to my attention.

@Eblo if you'd like to give this one last test run today, feel free.

Splendide-Imaginarius commented:
Thanks @WaywardHeart for your awesome work! Sorry it took so long to get this merged.
